README.md
# fuse.experiment

FIXME: my new application.

## Installation

Download from https://github.com/daveduthie/fuse.experiment

## Usage

FIXME: explanation

Run the project directly, via `:exec-fn`:

    $ clojure -X:run-x
    Hello, Clojure!

Run the project, overriding the name to be greeted:

    $ clojure -X:run-x :name '"Someone"'
    Hello, Someone!

Run the project directly, via `:main-opts` (`-m daveduthie.fuse.experiment`):

    $ clojure -M:run-m
    Hello, World!

Run the project, overriding the name to be greeted:

    $ clojure -M:run-m Via-Main
    Hello, Via-Main!

Run the project's tests (they'll fail until you edit them):

    $ clojure -M:test:runner

Build an uberjar:

    $ clojure -X:uberjar

This will update the generated `pom.xml` file to keep the dependencies synchronized with
your `deps.edn` file. You can update the version (and SCM tag) information in the `pom.xml` using the
`:version` argument:

    $ clojure -X:uberjar :version '"1.2.3"'

If you don't want the `pom.xml` file in your project, you can remove it, but you will
also need to remove `:sync-pom true` from the `deps.edn` file (in the `:exec-args` for `depstar`).

Run that uberjar:

    $ java -jar fuse.experiment.jar

## Options

FIXME: listing of options this app accepts.

## Examples

...

### Bugs

...

### Any Other Sections
### That You Think
### Might be Useful

## License

Copyright © 2021 Dave

_EPLv1.0 is just the default for projects generated by `clj-new`: you are not_
_required to open source this project, nor are you required to use EPLv1.0!_
_Feel free to remove or change the `LICENSE` file and remove or update this_
_section of the `README.md` file!_

Distributed under the Eclipse Public License version 1.0.


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [Unreleased]
### Changed
- Add a new arity to `make-widget-async` to provide a different widget shape.

## [0.1.1] - 2021-04-24
### Changed
- Documentation on how to make the widgets.

### Removed
- `make-widget-sync` - we're all async, all the time.

### Fixed
- Fixed widget maker to keep working when daylight savings switches over.

## 0.1.0 - 2021-04-24
### Added
- Files from the new template.
- Widget maker public API - `make-widget-sync`.

[Unreleased]: https://github.com/daveduthie/fuse.experiment/compare/0.1.1...HEAD
[0.1.1]: https://github.com/daveduthie/fuse.experiment/compare/0.1.0...0.1.1


test/daveduthie/fuse/experiment_test.clj
(ns daveduthie.fuse.experiment-test
  (:require [clojure.test :refer :all]
            [daveduthie.fuse.experiment :refer :all]))

(deftest a-test
  (testing "FIXME, I fail."
    (is (= 0 1))))


src/daveduthie/fuse/Experiment.java
package daveduthie.fuse;

import jnr.ffi.Platform;
import jnr.ffi.Pointer;
import jnr.ffi.types.off_t;
import jnr.ffi.types.size_t;
import ru.serce.jnrfuse.ErrorCodes;
import ru.serce.jnrfuse.FuseFillDir;
import ru.serce.jnrfuse.FuseStubFS;
import ru.serce.jnrfuse.struct.FileStat;
import ru.serce.jnrfuse.struct.FuseFileInfo;

import java.nio.file.Paths;
import java.util.Objects;

/**
 * @author Sergey Tselovalnikov
 * @see <a href="http://fuse.sourceforge.net/helloworld.html">helloworld</a>
 * @since 31.05.15
 */
public class Experiment extends FuseStubFS {

    public static final String HELLO_PATH = "/hello";
    public static final String HELLO_STR = "Hello World!";

    @Override
    public int getattr(String path, FileStat stat) {
        int res = 0;
        if (Objects.equals(path, "/")) {
            stat.st_mode.set(FileStat.S_IFDIR | 0755);
            stat.st_nlink.set(2);
        } else if (HELLO_PATH.equals(path)) {
            stat.st_mode.set(FileStat.S_IFREG | 0444);
            stat.st_nlink.set(1);
            stat.st_size.set(HELLO_STR.getBytes().length);
        } else {
            res = -ErrorCodes.ENOENT();
        }
        return res;
    }

    @Override
    public int readdir(String path, Pointer buf, FuseFillDir filter, @off_t long offset, FuseFileInfo fi) {
        if (!"/".equals(path)) {
            return -ErrorCodes.ENOENT();
        }

        filter.apply(buf, ".", null, 0);
        filter.apply(buf, "..", null, 0);
        filter.apply(buf, HELLO_PATH.substring(1), null, 0);
        return 0;
    }

    @Override
    public int open(String path, FuseFileInfo fi) {
        if (!HELLO_PATH.equals(path)) {
            return -ErrorCodes.ENOENT();
        }
        return 0;
    }

    @Override
    public int read(String path, Pointer buf, @size_t long size, @off_t long offset, FuseFileInfo fi) {
        if (!HELLO_PATH.equals(path)) {
            return -ErrorCodes.ENOENT();
        }

        byte[] bytes = HELLO_STR.getBytes();
        int length = bytes.length;
        if (offset < length) {
            if (offset + size > length) {
                size = length - offset;
            }
            buf.put(0, bytes, 0, bytes.length);
        } else {
            size = 0;
        }
        return (int) size;
    }

    public static void main(String[] args) {
        HelloFuse stub = new HelloFuse();
        try {
            String path;
            switch (Platform.getNativePlatform().getOS()) {
                case WINDOWS:
                    path = "J:\\";
                    break;
                default:
                    path = "/tmp/mnth";
            }
            stub.mount(Paths.get(path), true, true);
        } finally {
            stub.umount();
        }
    }
}


src/daveduthie/fuse/experiment.clj
(ns daveduthie.fuse.experiment
  (:import
   (java.io File)
   (java.nio ByteBuffer)
   (java.nio.file Paths)
   (jnr.ffi Pointer)
   (ru.serce.jnrfuse ErrorCodes FuseStubFS FuseFillDir)
   (ru.serce.jnrfuse.struct FileStat FuseFileInfo)))

(defn enoent-error []
  (* -1 (ErrorCodes/ENOENT)))

(def HELLO_STR "Hello World!")
(def HELLO_PATH "/hello")

(def LOGGER (agent nil))
(defn log [msg] (send-off LOGGER (fn [_] (println (pr-str msg)) nil)))

(defn fuse-custom-mount
  []
  (proxy [FuseStubFS] []
    (getattr [^String path ^FileStat stat]
      (log {:in :getattr :path path})
      (cond
        (= path "/")      (do (-> stat .-st_mode (.set (bit-or FileStat/S_IFDIR 0755)))
                              (-> stat .-st_nlink (.set 2))
                              0)
        (= path "/hello") (do (-> stat .-st_mode (.set (bit-or FileStat/S_IFREG 0444)))
                              (-> stat .-st_nlink (.set 1))
                              (-> stat .-st_size (.set (count (.getBytes HELLO_STR))))
                              0)
        :else             (enoent-error)))
    (readdir [^String path
              ^Pointer buf
              ^FuseFillDir filt
              ^Long offset
              ^FuseFileInfo fi]
      ;; Here we choose what to list.
      (log {:in :readdir})
      (if (= "/" path)
        (do (.apply filt buf "." nil 0)
            (.apply filt buf ".." nil 0)
            (.apply filt buf (.substring HELLO_PATH 1) nil 0)
            0)
        (enoent-error)))
    (open [^String path ^FuseFileInfo fi]
      ;; Here we handle errors on opening
      (log {:in :open :path path})
      (if (= "/hello" path) 0 (enoent-error)))
    (read [^String path ^Pointer buf ^Long size ^Long offset ^FuseFileInfo fi]
      ;; Here we read the contents
      (log {:in :read :path path})
      (if-not (= "/hello" path)
        (enoent-error)
        (let [bytes      (.getBytes HELLO_STR)
              len        (count bytes)
              to-read    (min (- len offset))
              contents   (ByteBuffer/wrap bytes)
              bytes-read (byte-array to-read)]
          (doto contents (.position offset) (.get bytes-read 0 to-read))
          (-> buf
              (.put 0 bytes-read 0 to-read))
          (.position contents 0)
          to-read)))))

(def stub-atom (atom nil))

(defn string-to-uri [s]
  (-> s File. .toURI))

(defn uri-to-path [s]
  (Paths/get s))

(defn string-to-path [s]
  (-> s string-to-uri uri-to-path))

(defn mount-it!
  [dir]
  (let [stub (fuse-custom-mount)]
    (log {:stub stub})
    (future (reset! stub-atom stub)
            ;; params: path blocking debug options
            (->
              stub
              (.mount
               (string-to-path dir)
               true
               true
               (into-array String []))))))

(comment
  (string-to-path "/tmp/fuse-experiments3"))

(defn umount-it! []
  (-> @stub-atom .umount))

(defn cleanup-hooks [mnt]
  (.addShutdownHook
   (Runtime/getRuntime)
   (Thread. (fn []
              (println "Unmounting " mnt)
              (umount-it!)))))

(defn main
  "I don't do a whole lot ... yet."
  [{:keys [dir]}]
  (cleanup-hooks dir)
  (println "Mounting " dir)
  (println (format "Try going to %s and running ls." dir))
  (deref (mount-it! dir)))

(comment
  (def DIR "/tmp/fuse-experiment2")
  (cleanup-hooks DIR)
  (deref (mount-it! DIR))
  @stub-atom
  (umount-it!)
  ,)
